### ISSUE TITLE
Properties for defining macros or variables

### ISSUE THREAD

**Issue by utelle** [premake/premake-core#551]
For a Visual Studio project I have a `.props` file that defines certain configuration macros that are used at other places using the `$(macroname)` syntax.

I found premake's `externalrule` for including external property files. However, this not only tries to include the external `.props` file, but also a `.target` file (which doesn't exist in my case). And it seems to be bound to handling files of a certain type.

Is there a way to generate the necessary `import` syntax for an external `.props` file containing generic macros? If yes, how? If no, is it planned to add such a feature to premake?

A similar mechanism would be useful for generated makefiles (gmake). In makefiles you can define variables like `FOO ?= foo` and can override the value of `FOO` by adding something like `FOO=bar` as a command line parameter on invoking make. Is this supported by premake? If yes, how? If no, is it planned to add such a feature?

**Comment by starkos** [comment]
Perhaps we need an `externalprops()` call? Though we're starting to get a bunch of those; I wonder if we might come up with a more generic function that can look at the linked file's extension and figure out what to do with it.

**Comment by utelle** [comment]
An `externalprops()` call would certainly be nice to have. Another (or additional) approach could be to allow to define such general properties with default values, and to generate a `.props` file for Visual Studio (or to add appropriate definition lines at the beginning of a makefile).

**Comment by tvandijck** [comment]
We had prop files like that too... and then we realized that premake lua scripts are effectively the same, so we took all those props, made proper premake scripts from it, and used them as premake tokens instead. Turned out a lot cleaner, and more importantly, all our 'props' now suddenly worked on xcode, gmake, and any other backend on any other OS.

Not that I'm against an externalprops or anything like that... but in all honesty, if you 'embraced the premake', you might not actually need it at all.. ;)

**Comment by starkos** [comment]
> Not that I'm against an externalprops or anything like that... but in all honesty, if you 'embraced the premake', you might not actually need it at all.. ;)

The case for including it is when you need to pre-generate projects which will then distribute to other groups, who will then only use Visual Studio. That was also why we added the Premake-token-to-VS-magic-symbol feature: so downstream users could tweak the generated VS project settings easily, without needing Premake.

**Comment by utelle** [comment]
@tvandijck: I agree that it could be done in premake itself. In fact, that's how I currently do it. However, one big advantage of premake is that build files can be pre-generated (as @starkos pointed out), and that spares the users of my components to struggle with yet another build tool.

The users of my wxWidgets components are used to get pre-generated build files. However, I would like to make it easy to adjust certain configuration options (like the version number of the library against which the components will be compiled). And to edit a single small property file for Visual Studio (or adding a few command line parameters for make) seems to be appropriate.

**Comment by stolk** [comment]
SDKs for game consoles can come with .prop files so it would be nice to be able to just refer to those.

As that's easier than trying to replicate them with premake, and then get out-dated when the SDK is updated?
